editor/src/clj/editor/sprite.clj

```diff
@@ -453,91 +496,111 @@
-                                          attribute-prop {:node-id _node-id
-                                                          :type attribute-type
-                                                          :edit-type attribute-edit-type
-                                                          :value (fill-with-zeros attribute-values attribute-expected-value-count)
-                                                          :label (properties/keyword->name attribute-key)}]
-                                      ;; Insert the original material values as original-value if there is a vertex override
-                                      (if (contains? vertex-attribute-overrides attribute-key)
-                                        {attribute-property-key (assoc attribute-prop :original-value (:values attribute))}
-                                        {attribute-property-key attribute-prop})))
-                                  material-attribute-infos)]
+  (let [attribute-properties
+        (keep (fn [attribute-info]
+                (when (editable-attribute-info? attribute-info)
+                  (let [attribute-key (:name-key attribute-info)
+                        semantic-type (:semantic-type attribute-info)
+                        material-values (:values attribute-info)
+                        override-values (vertex-attribute-overrides attribute-key)
+                        attribute-values (or override-values material-values)
+                        property-type (attribute-property-type attribute-info)
+                        expected-element-count (attribute-expected-element-count attribute-info)
+                        edit-type (attribute-edit-type attribute-info property-type)
+                        property-key (attribute-key->property-key attribute-key)
+                        label (properties/keyword->name attribute-key)
+                        value (if (= g/Num property-type)
+                                (first attribute-values) ; The widget expects a number, not a vector.
+                                (graphics/resize-doubles attribute-values semantic-type expected-element-count))
+                        error (when (some? override-values)
+                                (graphics/validate-doubles override-values attribute-info _node-id property-key))
+                        prop {:node-id _node-id
+                              :type property-type
+                              :edit-type edit-type
+                              :label label
+                              :value value
+                              :error error}]
+                    ;; Insert the original material values as original-value if there is a vertex override.
+                    (if (some? override-values)
+                      [property-key (assoc prop :original-value material-values)]
+                      [property-key prop]))))
+              material-attribute-infos)]
```

Comment:
Mostly whitespace + renames, but we also decorate invalid attributes with errors to make them red and get an informative tooltip in the property editor.